File: 3_Stack/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        # create hashmap to open parentheses
        Map = {")": "(", "]": "[", "}": "{"}
        stack = []

        # check if char is in hashmap
        for c in s:
            # if NOT a closing parentheses
            if c not in Map:
                stack.append(c)    # add to stack
                continue
            
            # if NOT in stack OR NOT closing parentheses 
            if not stack or stack[-1] != Map[c]:
                return False
            
            # if above conditions are not met then parentheses must match
            stack.pop()

        # returns true if the stack is empty 
        return not stack

# implementation
answer = Solution()

print("Leetcode 20: Valid Parentheses\n")

# Example 1
example_1 = "()"
print(f"Example 1: s = {example_1}")
print("Output:", answer.isValid(example_1))

# Example 2
example_2 = "()[]{}"
print(""f"Example 3: s = {example_2}")
print("Output:", answer.isValid(example_2))

# Example 3 
example_3 = "(]"
print(f"Example 3: s = {example_3}")
print("Output:", answer.isValid(example_3))

File: 03_Stack/validParentheses.js
var isValid = function(s) {
    // create hashmap
    const hashmap = {'}':'{', ']':'[', ')':'('};
    // create stack
    let stack = [];
    
    // check if char is in hashmap
    for (let index = 0; index < s.length; index++) {
        let char = s[index];
        
        // check if NOT closing paretheses
        if (!(char in hashmap)) {
            stack.push(char);
            continue;
        }
        
        // if NOT in stack OR NOT closing parethesis
        if (stack.length === 0 || stack[stack.length- 1] !== hashmap[char]) {
            return false;
        }
        
        // if the above positions are not met then we have a match!
        stack.pop();
    }
    // returns true is the stack is empty
    return stack.length === 0;
};

// implementation

console.log("Leetcode 20: Valid Parentheses\n");

// Example 1
const example1 = "()";
console.log(`Example 1: s = ${example1}`);
console.log("Output", isValid(example1));

// Example 2
const example2 = "()[]{}";
console.log(`Example 2: s = ${example2}`);
console.log("Output:", isValid(example2));

// Example 3
const example3 = "(]";
console.log(`Example 3: s = ${example3}`);
console.log("Output:", isValid(example3));